### ISSUE TITLE
Single location for read/write data files in egeria

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#1102]
One obervation that applies across egeria -- we really need to ensure that, without too much config, all data (writing) is constrained to a single/small number of folders. this is essential when working with containers which are basically read-only, but with specific read-write volumes mounted at a specific location. Each volume has overhead - sometimes sys admin created, so having files in many places would be tricky.

I will raise a seperate issue to track this, since it's much broader than just lineage - which is only following the existing pattern! .. and there isn't really precedent set yet.

_Originally posted by @planetf1 in https://github.com/odpi/egeria/pull/1100_

**Comment by planetf1** [comment]
Raised in reference to a review here: https://github.com/odpi/egeria/pull/1101/files

**Comment by mandy-chessell** [comment]
I do not think this is a concern - locations where data is written to is controlled by the connectors.  Is there a problem with any of the defaults?

**Comment by planetf1** [comment]
will review current locations

**Comment by planetf1** [comment]
cc: @mandy-chessell

**Comment by planetf1** [comment]
still valid

**Comment by planetf1** [comment]
This has come up again whilst looking at #4302 and affects any data the server chassis process writes to disk.

This could include
 - log files
 - configuration documents
  - local graph repo (BSD)

In order to be persisted, in docker (and via the container in k8s etc too), we need to setup volumes (docker) or pvcs (k8s) at 'deployment' time to manage any persistent data the container might store. 

To do this we specify a virtual location in the image, and map this to a disk/volume managed by the infrastructure. So we need to know what that location is.

In egeria the above functionality makes use of connectors. For example the configuration file connector can be configured with a path to use - this works fine, but can only be issued once the platform is active, and must be the same as that specified by the person doing the container deployment to work properly. We default to 'current working directory' of the server chassis process.

In the case of the graph repo, we can specify janusgraph properties when the egeria *server* is configured. We provide an on-disk location - so in docker/k8s the path must match to work correctly. The default is again based off the server chassi's current working directory

-- not checked log etc connectors yet.

It would be a lot easier if, by default, at a minimum we designated a clear directory that is used for persisted data ie we might use
  - data/config
  - data/repos
  - data/logs

Indeed in some container environments we can expect anything OTHER than what we specify to be read-only (something that should be tested)

this way we could easily default files to be stored in those places in the code, and also in our docker image/k8s environments create the right volumes by default.

The existing APIs could still be used for other scenarios, but we'd have something predictable and simple that would 'just work' in most cases.

We could of course manage this configuration in k8s via initialization scripts whereby we might run a client against the server immediately after startup which, for example, would issue the REST APi call to change the config directory -- but how do we prevent user apps from changing/breaking this? Or issueing config requests too soon? and what about when creating a metadata server repo? Does that user know the right filepath to use? what about the server author UI?

This being said, as we progress - for example in operator work - to create servers driven by CRDs, we will likely need an 'admin platform' active that can create the required config documents so I suspect this is an area we will keep coming back to - or have some other way of loading/enforcing config once a platform is launched.

**Comment by planetf1** [comment]
cc: @mandy-chessell @grahamwallis @davidradl

**Comment by planetf1** [comment]
One option is to allow the default to be specified via a jvm or spring property, that could avoid impacting existing egeria users with a change, but could be set by default in the docker/container environment - if we were to do this I would probably suggest a single parameter to be used as the 'root' - instead of 'current directory', but also need to evaluate where ANY code in egeria writes to and document, and anticipate that being a small number of directory trees

**Comment by planetf1** [comment]
I've changed the default locations as follows:

Configuration: ./data/config/omag.server.{0}.config (0=server name)
Graph repo: ./data/repository/{0} (0 = serverRepositoryPath)
Cohort: ./data/cohort/{0}.{1}.registrystore (0=server name, 1 = cohort name)

Lineage: Added a note to the readme recommending best practice in container environment. There are no defaults and the locations are always configured by hand. Noted the examples don't allow for multiple servers on same platform, so added that to recommendation also

In each case the principle is to use './data' (relative to current working directory) as the location for all writeable data, which could then easily be a mountable volume. we could split deeper if needed -

I will test with this change and would be sufficient for immediate ongoing work

HOWEVER
This brings up an interesting question.

Would it be better to have
 ./data/servers/cocoMDS1/config
 ./data/servers/cocoMDS1/cohorts
 ./data/servers/cocoMDS1/repository
 ./data/server/cocoMDS1/lineage
 
 as the top levels.
 
 This way we:
  a) Have all dynamic data under 'data'
  b) Keep all config relating to a single server in a single directory -- this could also later be mounted, and could be easier as we continue the operator work which may dynamically create servers
  c) Names the subdirectories more clearly so we know what is in each (we could keep file names as-is, or shortern given the context is now in the path)
  d) Allows for platform level config in future (ie under ./data/platform etc as needed... )
  
  This is probably cleaner and more logical than my first iteration. Thoughts @mandy-chessell

**Comment by planetf1** [comment]
The latter suggestion above is also clear in going from most general to most specific ie egeria / dynamic data / server info / myserver / type of data

**Comment by planetf1** [comment]
Having tested the change against the coco pharam scenario (config, start, data catalog) I have checked
 - no new files are written anywhere in the tree except 'data'
 - tree looks like as follows
 
 ```
 jonesn:egeria/ (issue1102*) $ cat /tmp/t | grep -v vertex | grep -v edge                                                               [13:27:56]
./data
./data/servers
./data/servers/cocoMDS3-graph-repository
./data/servers/cocoMDS3-graph-repository/repository
./data/servers/cocoMDS3-graph-repository/repository/graph
./data/servers/cocoMDS3-graph-repository/repository/graph/berkeley
./data/servers/cocoMDS3-graph-repository/repository/graph/berkeley/00000000.jdb
./data/servers/cocoMDS3-graph-repository/repository/graph/berkeley/je.config.csv
./data/servers/cocoMDS3-graph-repository/repository/graph/berkeley/je.stat.csv
./data/servers/cocoMDS3-graph-repository/repository/graph/berkeley/je.info.0.lck
./data/servers/cocoMDS3-graph-repository/repository/graph/berkeley/je.lck
./data/servers/cocoMDS3-graph-repository/repository/graph/berkeley/je.info.0
./data/servers/cocoMDS3-graph-repository/repository/graph/searchindex
./data/servers/cocoView1
./data/servers/cocoView1/config
./data/servers/cocoView1/config/cocoView1.config
./data/servers/cocoMDS6
./data/servers/cocoMDS6/config
./data/servers/cocoMDS6/config/cocoMDS6.config
./data/servers/cocoMDS6/cohorts
./data/servers/cocoMDS6/cohorts/iotCohort.registrystore
./data/servers/cocoMDS6/cohorts/cocoCohort.registrystore
./data/servers/cocoMDS1
./data/servers/cocoMDS1/config
./data/servers/cocoMDS1/config/cocoMDS1.config
./data/servers/cocoMDS1/cohorts
./data/servers/cocoMDS1/cohorts/cocoCohort.registrystore
./data/servers/governDL01
./data/servers/governDL01/config
./data/servers/governDL01/config/governDL01.config
./data/servers/cocoMDS2-graph-repository
./data/servers/cocoMDS2-graph-repository/repository
./data/servers/cocoMDS2-graph-repository/repository/graph
./data/servers/cocoMDS2-graph-repository/repository/graph/berkeley
./data/servers/cocoMDS2-graph-repository/repository/graph/berkeley/00000000.jdb
./data/servers/cocoMDS2-graph-repository/repository/graph/berkeley/je.config.csv
./data/servers/cocoMDS2-graph-repository/repository/graph/berkeley/je.stat.csv
./data/servers/cocoMDS2-graph-repository/repository/graph/berkeley/je.info.0.lck
./data/servers/cocoMDS2-graph-repository/repository/graph/berkeley/je.lck
./data/servers/cocoMDS2-graph-repository/repository/graph/berkeley/je.info.0
./data/servers/cocoMDS2-graph-repository/repository/graph/searchindex
./data/servers/cocoMDS6-graph-repository
./data/servers/cocoMDS6-graph-repository/repository
./data/servers/cocoMDS6-graph-repository/repository/graph
./data/servers/cocoMDS6-graph-repository/repository/graph/berkeley
./data/servers/cocoMDS6-graph-repository/repository/graph/berkeley/00000000.jdb
./data/servers/cocoMDS6-graph-repository/repository/graph/berkeley/je.config.csv
./data/servers/cocoMDS6-graph-repository/repository/graph/berkeley/je.stat.csv
./data/servers/cocoMDS6-graph-repository/repository/graph/berkeley/je.info.0.lck
./data/servers/cocoMDS6-graph-repository/repository/graph/berkeley/je.lck
./data/servers/cocoMDS6-graph-repository/repository/graph/berkeley/je.info.0
./data/servers/cocoMDS6-graph-repository/repository/graph/searchindex
./data/servers/cocoMDS1-graph-repository
./data/servers/cocoMDS1-graph-repository/repository
./data/servers/cocoMDS1-graph-repository/repository/graph
./data/servers/cocoMDS1-graph-repository/repository/graph/berkeley
./data/servers/cocoMDS1-graph-repository/repository/graph/berkeley/00000000.jdb
./data/servers/cocoMDS1-graph-repository/repository/graph/berkeley/je.config.csv
./data/servers/cocoMDS1-graph-repository/repository/graph/berkeley/je.stat.csv
./data/servers/cocoMDS1-graph-repository/repository/graph/berkeley/je.info.0.lck
./data/servers/cocoMDS1-graph-repository/repository/graph/berkeley/je.lck
./data/servers/cocoMDS1-graph-repository/repository/graph/berkeley/je.info.0
./data/servers/cocoMDS1-graph-repository/repository/graph/searchindex
./data/servers/cocoMDSx-graph-repository
./data/servers/cocoMDSx-graph-repository/repository
./data/servers/cocoMDSx-graph-repository/repository/graph
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley/00000000.jdb
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley/je.config.csv
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley/je.stat.csv
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley/je.info.0.lck
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley/je.lck
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley/je.info.0
./data/servers/cocoMDSx-graph-repository/repository/graph/searchindex
./data/servers/cocoMDSx
./data/servers/cocoMDSx/config
./data/servers/cocoMDSx/config/cocoMDSx.config
./data/servers/cocoMDSx/cohorts
./data/servers/cocoMDSx/cohorts/devCohort.registrystore
./data/servers/cocoMDS4
./data/servers/cocoMDS4/config
./data/servers/cocoMDS4/config/cocoMDS4.config
./data/servers/cocoMDS4/cohorts
./data/servers/cocoMDS4/cohorts/cocoCohort.registrystore
./data/servers/cocoMDS3
./data/servers/cocoMDS3/config
./data/servers/cocoMDS3/config/cocoMDS3.config
./data/servers/cocoMDS3/cohorts
./data/servers/cocoMDS3/cohorts/cocoCohort.registrystore
./data/servers/cocoMDS2
./data/servers/cocoMDS2/config
./data/servers/cocoMDS2/config/cocoMDS2.config
./data/servers/cocoMDS2/cohorts
./data/servers/cocoMDS2/cohorts/devCohort.registrystore
./data/servers/cocoMDS2/cohorts/iotCohort.registrystore
./data/servers/cocoMDS2/cohorts/cocoCohort.registrystore
./data/servers/cocoMDS5
./data/servers/cocoMDS5/config
./data/servers/cocoMDS5/config/cocoMDS5.config
./data/servers/cocoMDS5/cohorts
./data/servers/cocoMDS5/cohorts/cocoCohort.registrystore
```

In addition we have lots of files from the graph store like
```
./data/servers/cocoMDS2-graph-repository/repository/graph/searchindex/edgeIndexPropertyermetadataCollectionName/_c.cfs
```
or
```
./data/servers/cocoMDS1-graph-repository/repository/graph/searchindex/vertexIndexPropertyvemetadataCollectionName/_10.si
```
etc

I have NOT tested OLS, since this is not covered in the demo scenario environment  

Additionally I have configured the default paths for auditlog as `        String endpointAddress = "data/servers/" + localServerName + "logs/auditlog";
` but though these appear to be configured in the notebooks, I am not seeing audit logs - will debug

So in summary files look like:
```
./data/servers/cocoMDS5/cohorts/cocoCohort.registrystore
./data/servers/cocoMDS5/config/cocoMDS5.config
./data/servers/cocoMDSx-graph-repository/repository/graph/berkeley
./data/servers/cocoMDSx-graph-repository/repository/graph/searchindex
```

* We could extend the cohort path to be cohorts/cocoCohort/cocoCohort.registrystore ? 
* we could drop the server/cohort name from the filename in the first two example - but does this make it harder to save the files somewhere safe?
* There is no migration or fall-back to old filenames
* debugging the audit log connector

**Comment by planetf1** [comment]
Audit logs are now tested by
 * running the current coco pharma 'config' notebook
 
 ```
http --verbose --verify=no --pretty=format DELETE https://localhost:9443/open-metadata/admin-services/users/garygeeke/servers/cocoMDS2/audit-log-destinations
echo '[ ]' | http --verbose --verify=no --pretty=format POST https://localhost:9443/open-metadata/admin-services/users/garygeeke/servers/cocoMDS2/audit-log-destinations/files
http --verbose --verify=no --pretty=format DELETE https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoMDS1/audit-log-destinations
http --verbose --verify=no --pretty=format DELETE https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoMDS4/audit-log-destinations
echo '[ ]' | http --verbose --verify=no --pretty=format POST https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoMDS1/audit-log-destinations/files
echo '[ ]' | http --verbose --verify=no --pretty=format POST https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoMDS4/audit-log-destinations/files
```
Then running the 'start' notebook (which results in plenty of audit log messages).

We observe less audit log events on stdout, since some are going to file (cocoMDS 1/2/4) & observe a directory structure like:
```
jonesn:auditlog/ (issue1102) $ pwd                                                                                                   
/Users/jonesn/src/egeria/data/servers/cocoMDS1/logs/auditlog
jonesn:auditlog/ (issue1102) $ ls | head -5                                                                                         
log-record-001f7dcb-4d97-4794-9085-44bbc0b9eaf8
log-record-01046306-0009-4c10-89fc-885f4a62e8f5
log-record-0143204d-58c2-4702-ac8a-dbb76a1886c4
log-record-01693fb0-c0aa-4e3d-9857-f6db2a4c2957
log-record-01a4afb8-30a1-40fb-84b0-71041f86419d
```
This is as expected & documented

I have now moved the PR to review state & welcome comments.

Have any files been missed?

Should now be able to continue with #4302

**Comment by grahamwallis** [comment]
I can't think of any missing files.

**Comment by planetf1** [comment]
Note that for the graph repo the directory no longer contains '-graph-repository'. I pushed a fix yesterday evening having noticed I'd left that qualifier in --
